parquet/src/test/java/org/apache/iceberg/parquet/TestVariantReaders.java

```diff
@@ -0,0 +1,1107 @@
+    MessageType parquetSchema = parquetSchema(variantType);
+
+    GenericRecord variant =
+        record(variantType, Map.of("metadata", VariantTestUtil.emptyMetadata()));
+    GenericRecord record = record(parquetSchema, Map.of("id", 1, "var", variant));
+
+    Record actual = writeAndRead(parquetSchema, record);
+    assertThat(actual.getField("id")).isEqualTo(1);
+    assertThat(actual.getField("var")).isInstanceOf(Variant.class);
+
+    Variant actualVariant = (Variant) actual.getField("var");
+    VariantTestUtil.assertEqual(EMPTY_METADATA, actualVariant.metadata());
+    VariantTestUtil.assertEqual(Variants.ofNull(), actualVariant.value());
+  }
+
+  @Test
+  public void testMissingValueColumn() throws IOException {
+    GroupType variantType =
+        Types.buildGroup(Type.Repetition.REQUIRED)
+            .id(2)
+            .required(PrimitiveTypeName.BINARY)
+            .named("metadata")
+            .addField(shreddedPrimitive(PrimitiveTypeName.INT32))
+            .named("var");
+    MessageType parquetSchema = parquetSchema(variantType);
+
+    GenericRecord variant =
+        record(variantType, Map.of("metadata", VariantTestUtil.emptyMetadata(), "typed_value", 34));
+    GenericRecord record = record(parquetSchema, Map.of("id", 1, "var", variant));
+
+    Record actual = writeAndRead(parquetSchema, record);
+    assertThat(actual.getField("id")).isEqualTo(1);
+    assertThat(actual.getField("var")).isInstanceOf(Variant.class);
+
+    Variant actualVariant = (Variant) actual.getField("var");
+    VariantTestUtil.assertEqual(EMPTY_METADATA, actualVariant.metadata());
+    VariantTestUtil.assertEqual(Variants.of(34), actualVariant.value());
+  }
+
+  @Test
+  public void testValueAndTypedValueConflict() throws IOException {
+    GroupType variantType = variant("var", 2, shreddedPrimitive(PrimitiveTypeName.INT32));
+    MessageType parquetSchema = parquetSchema(variantType);
+
+    GenericRecord variant =
+        record(
+            variantType,
+            Map.of(
+                "metadata",
+                VariantTestUtil.emptyMetadata(),
+                "value",
+                serialize(Variants.of("str")),
+                "typed_value",
+                34));
+    GenericRecord record = record(parquetSchema, Map.of("id", 1, "var", variant));
+
+    assertThatThrownBy(() -> writeAndRead(parquetSchema, record))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("Invalid variant, conflicting value and typed_value");
+  }
+
+  @Test
+  public void testUnsignedInteger() {
+    GroupType variantType =
+        variant(
+            "var",
+            2,
+            shreddedPrimitive(PrimitiveTypeName.INT32, LogicalTypeAnnotation.intType(32, false)));
+    MessageType parquetSchema = parquetSchema(variantType);
+
+    GenericRecord variant =
+        record(variantType, Map.of("metadata", VariantTestUtil.emptyMetadata()));
```

Comment:
Addd typed_value 
`record(variantType, Map.of("metadata", VariantTestUtil.emptyMetadata(), "typed_value", 34));` to have the shredded value?

Reply:
We don't need to. This fails before attempting to read because the schema is invalid.